tests/fx/parallelization/test_tensor_parallel.py

```diff
@@ -0,0 +1,190 @@
+        "position_ids": torch.arange(0, 10, device=device).unsqueeze(0),
+    }
+    
+    # different shape to trigger recompile
+    another_inputs = {
+        "input_ids": torch.randint(low=1, high=model_config.vocab_size, size=(1, 11), device=device),
+        "attention_mask": torch.ones((1, 11), dtype=torch.int64, device=device),
+        "position_ids": torch.arange(0, 11, device=device).unsqueeze(0),
+    }
+
+    # this will initialize all linears on meta device
+    with MetaAwareMethodsPatcher():
+        model = model_cls(model_config)
+        model.eval()
+    # move model to current device, with linears still on meta
+    move_model_to_device(model, device=device)
+    initialize_parameter_mapping(model, ctx=ctx)
+
+    model = torch.compile(model, fullgraph=True, backend=partial(parallelize_backend, ctx=ctx, config=cfg))
+    model(**inputs)
+
+    parameter_ids = set([id(param) for _, param in model.named_parameters()])
+    model(**another_inputs)
+
+    parameter_ids_after_recompile = set([id(param) for _, param in model.named_parameters()])
+    assert parameter_ids == parameter_ids_after_recompile
+
+    dist.barrier(tp_group)
+    tearDown(tp_group)
+
+def run_test_parallel_results_matches_non_parallel(rank: int, world_size: int, model_cls: Type[PreTrainedModel], model_config: PretrainedConfig):
+    dummify(model_config)
+
+    dist_init(rank, world_size)
+    tp_group = dist.new_group(ranks=[rank])
+    
+    # prepare config and context
+    device = torch.device(type='cuda', index=torch.cuda.current_device())
+    ctx, cfg = ParallelExecutionCtx(tp_group=tp_group, current_device=device), Config()
+    
+    inputs = {
+        "input_ids": torch.randint(low=1, high=model_config.vocab_size, size=(1, 10), device=device),
+        "attention_mask": torch.ones((1, 10), dtype=torch.int64, device=device),
+        "position_ids": torch.arange(0, 10, device=device).unsqueeze(0),
+    }
+
+    set_seed(SEED)
+    # non-parallel local forward
+    with MetaAwareMethodsPatcher():
+        model = model_cls(model_config)
+        model.eval()
+
+    # move model to current device, with linears still on meta
+    move_model_to_device(model, device=device)
+    initialize_parameter_mapping(model, ctx=ctx)
+
+    model = torch.compile(model, fullgraph=True, backend=partial(parallelize_backend, ctx=ctx, config=cfg))
+    logits = model(**inputs)[0]
+
+    del model
+
+    tp_group = dist.new_group()
+    set_seed(SEED)
+    ctx = ParallelExecutionCtx(tp_group=tp_group, current_device=device)
+    with MetaAwareMethodsPatcher():
+        model = model_cls(model_config)
+        model.eval()
+
+    # move model to current device, with linears still on meta
+    move_model_to_device(model, device=device)
+    initialize_parameter_mapping(model, ctx=ctx)
+
+    model = torch.compile(model, fullgraph=True, backend=partial(parallelize_backend, ctx=ctx, config=cfg))
```

Comment:
can we compose with inductor?

Reply:
I would say that is the hope for the future.